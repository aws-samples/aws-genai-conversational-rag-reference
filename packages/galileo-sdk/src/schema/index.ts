/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
export enum ChainType {
  QA = 'QA',
  CONDENSE_QUESTION = 'CONDENSE_QUESTION',
  CLASSIFY = 'CLASSIFY',
}
