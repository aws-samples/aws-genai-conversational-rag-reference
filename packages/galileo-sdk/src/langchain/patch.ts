/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import { BaseMessage } from 'langchain/schema';

// Patch langchain BaseMessage prototype to have "type" getter
// Currently only has _getType() method and we need simple getter for template
Object.defineProperty(BaseMessage.prototype, 'type', {
  get: function () {
    return this._getType();
  },
  configurable: true,
  enumerable: true,
});
