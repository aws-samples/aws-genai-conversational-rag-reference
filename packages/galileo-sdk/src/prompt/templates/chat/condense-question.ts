/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import { BaseMessage } from 'langchain/schema';
import { BASE_CHAT_PARTIALS, ChatTemplatePartials } from './base.js';
import { HandlebarsPromptTemplate, HandlebarsPromptTemplateRuntime, ScopedHandlebarsPromptTemplateInput } from '../../handlebars.js';

export const CHAT_CONDENSE_QUESTION_PARTIALS: ChatTemplatePartials = {
  ...BASE_CHAT_PARTIALS,
  Body: '{{>Instruction}}\n\n{{>Dialog}}\n\n{{>Cue}}',
  Instruction: 'Given the following conversational dialog denoted by {{>Delimiter}}, and the "Followup Question" below, rephrase the "Followup Question" to be a concise standalone question in its original language. Without answering the question, return only the standalone question.',
  Cue: 'Followup Question: {{question}}\n\nStandalone Question: ',
} as const;

export interface ChatCondenseQuestionPromptTemplateInputValues {
  readonly chat_history: BaseMessage[];
  readonly question: string;
}

export type ChatCondenseQuestionPromptTemplateInput =
  ScopedHandlebarsPromptTemplateInput<ChatTemplatePartials, ChatCondenseQuestionPromptTemplateInputValues>;
export type ChatCondenseQuestionPromptRuntime =
  HandlebarsPromptTemplateRuntime<ChatCondenseQuestionPromptTemplateInput>;

export class ChatCondenseQuestionPromptTemplate extends
  HandlebarsPromptTemplate<ChatTemplatePartials, ChatCondenseQuestionPromptTemplateInputValues> {

  static async deserialize(data: any) {
    return new ChatCondenseQuestionPromptTemplate(data);
  }

  constructor(input: ChatCondenseQuestionPromptTemplateInput) {
    super({
      template: '{{>Layout}}',
      inputVariables: ['chat_history', 'question'],
      ...input,
      partialVariables: {
        ...input.partialVariables,
      },
      templatePartials: {
        ...CHAT_CONDENSE_QUESTION_PARTIALS,
        ...input.templatePartials,
      },
    });
  }
}
