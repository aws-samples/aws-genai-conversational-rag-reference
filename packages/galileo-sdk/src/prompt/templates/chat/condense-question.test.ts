/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import { AIMessage, HumanMessage, SystemMessage } from 'langchain/schema';
import { ChatCondenseQuestionPromptTemplate } from './condense-question.js';

describe('prompt/templates/chat', () => {
  describe('condense-question', () => {
    test('default', async () => {
      const template = new ChatCondenseQuestionPromptTemplate({});

      expect(await template.format({
        question: 'THE QUESTION',
        chat_history: [
          new SystemMessage('A system message'),
          new HumanMessage('Hi'),
          new AIMessage('Hello!'),
        ],
      })).toMatchSnapshot();
    });

    test('messages', async () => {
      const template = new ChatCondenseQuestionPromptTemplate({
        template: `
{{#*inline "HumanMessage"}}<|prompter|>{{content}}<|endoftext|>{{/inline}}
{{#*inline "AIMessage"}}<|assistant|>{{content}}<|endoftext|>{{/inline}}
{{#*inline "SystemMessage"}}<|system|>{{content}}<|endoftext|>{{/inline}}

{{>Layout}}
        `,
      });

      expect(await template.format({
        question: 'THE QUESTION',
        chat_history: [
          new SystemMessage('A system message'),
          new HumanMessage('Hi'),
          new AIMessage('Hello!'),
        ],
      })).toMatchSnapshot();
    });

    test('wrappers', async () => {
      const template = new ChatCondenseQuestionPromptTemplate({
        template: `
{{#*inline "Header"}}<SEQUENCE>{{/inline}}
{{#*inline "Footer"}}</SEQUENCE>{{/inline}}
{{#*inline "DialogHeader"}}<DIALOG>{{/inline}}
{{#*inline "DialogFooter"}}</DIALOG>{{/inline}}

{{>Layout}}
        `,
      });

      expect(await template.format({
        question: 'THE QUESTION',
        chat_history: [
          new SystemMessage('A system message'),
          new HumanMessage('Hi'),
          new AIMessage('Hello!'),
        ],
      })).toMatchSnapshot();
    });

    test('should allow complete override of template', async () => {
      const template = new ChatCondenseQuestionPromptTemplate({
        template: 'CUSTOM',
      });

      expect(await template.format({
        question: 'THE QUESTION',
        chat_history: [
          new SystemMessage('A system message'),
          new HumanMessage('Hi'),
          new AIMessage('Hello!'),
        ],
      })).toBe('CUSTOM');
    });
  });
});
