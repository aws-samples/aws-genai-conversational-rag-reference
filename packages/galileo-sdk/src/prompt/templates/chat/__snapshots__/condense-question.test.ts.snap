// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prompt/templates/chat condense-question default 1`] = `
"Given the following conversational dialog denoted by ''', and the "Followup Question" below, rephrase the "Followup Question" to be a concise standalone question in its original language. Without answering the question, return only the standalone question.
Dialog: '''
System: A system message

Human: Hi

Assistant: Hello!'''
Followup Question: THE QUESTION

Standalone Question: "
`;

exports[`prompt/templates/chat condense-question messages 1`] = `
"




Given the following conversational dialog denoted by ''', and the "Followup Question" below, rephrase the "Followup Question" to be a concise standalone question in its original language. Without answering the question, return only the standalone question.
Dialog: '''
<|system|>A system message<|endoftext|>

<|prompter|>Hi<|endoftext|>

<|assistant|>Hello!<|endoftext|>'''
Followup Question: THE QUESTION

Standalone Question:         "
`;

exports[`prompt/templates/chat condense-question wrappers 1`] = `
"





<SEQUENCE>Given the following conversational dialog denoted by ''', and the "Followup Question" below, rephrase the "Followup Question" to be a concise standalone question in its original language. Without answering the question, return only the standalone question.
Dialog: '''
System: A system message

Human: Hi

Assistant: Hello!'''
Followup Question: THE QUESTION

Standalone Question: </SEQUENCE>        "
`;
